### PR TITLE
session: migrate test-infra to testify for testSessionSerialSuite

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/28328

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/session_test.go [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+	"github.com/pingcap/tidb/session"
+	"github.com/pingcap/tidb/store/mockstore"
+	"github.com/pingcap/tidb/testkit"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/stretchr/testify/require"
+	"github.com/tikv/client-go/v2/tikv"
+	"github.com/tikv/client-go/v2/txnkv/transaction"
+)
+
+func TestSkipWithGrant(t *testing.T) {
```
Comment (tisonkun) [inline session/session_test.go:40]: We don't have to run these test with real tikv?

File: tests/realtikvtest/sessiontest/session_test.go [hunk 2]
```diff
@@ -1021,3 +1041,928 @@ func TestDatabase(t *testing.T) {
...
+
+	require.Nil(t, failpoint.Enable("github.com/pingcap/tidb/ddl/storeCloseInLoop", `return`))
+	defer failpoint.Disable("github.com/pingcap/tidb/ddl/storeCloseInLoop")
+
+	// this DDL call will enter deadloop before this fix
+	err = dom.DDL().CreateSchema(se, model.NewCIStr("testschema"), nil, nil)
+	require.Equal(t, "context canceled", err.Error())
+}
+
+func TestTemporaryTableInterceptor(t *testing.T) {
```
Comment (tisonkun) [inline tests/realtikvtest/sessiontest/session_test.go:1361]: This test doesn't belong to `testSessionSerialSuite` , would be covered by #34664, and you don't remove its counter part in session_legacy_test.go.

Please review all your changes to keep modification limited to `testSessionSerialSuite`.

File: tests/realtikvtest/sessiontest/session_test.go [hunk 3]
```diff
@@ -1021,3 +1041,928 @@ func TestDatabase(t *testing.T) {
...
+	}
+
+	// Also check GetSnapshotWithTS
+	snap := tk.Session().GetSnapshotWithTS(0)
+	val, err := snap.Get(context.Background(), k)
+	require.NoError(t, err)
+	require.Equal(t, []byte("v1"), val)
+}
+
+func TestLocalTemporaryTableUpdate(t *testing.T) {
```
Comment (tisonkun) [inline tests/realtikvtest/sessiontest/session_test.go:1417]: ditto

File: tests/realtikvtest/sessiontest/session_test.go [hunk 4]
```diff
@@ -1021,3 +1041,928 @@ func TestDatabase(t *testing.T) {
...
+		}
+		if sqlCase.additionalCheck != nil {
+			sqlCase.additionalCheck(err)
+		}
+		tk.MustExec("delete from tmp1")
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows())
+	}
+}
+
+func TestLocalTemporaryTableInsert(t *testing.T) {
```
Comment (tisonkun) [inline tests/realtikvtest/sessiontest/session_test.go:1608]: ditto

File: tests/realtikvtest/testkit.go [hunk 5]
```diff
@@ -52,6 +52,7 @@ func RunTestMain(m *testing.M) {
 	})
 	tikv.EnableFailpoints()
 	opts := []goleak.Option{
+		goleak.IgnoreTopFunction("github.com/pingcap/goleveldb/leveldb.(*DB).mpoolDrain"),
```
Comment (tisonkun) [inline tests/realtikvtest/testkit.go:55]: This leak can be resolved by:

```go
	callback := func(i int) int {
		// wait for MVCCLevelDB to close, MVCCLevelDB will be closed in one second
		time.Sleep(time.Second)
		testDataMap.GenerateOutputIfNeeded()
		return i
	}
	goleak.VerifyTestMain(testmain.WrapTestingM(m, callback), opts...)
```

**Review by tisonkun — APPROVED** [review]

File: tests/realtikvtest/testkit.go (hunk 5 above)
Comment (hawkingrei) [reply tests/realtikvtest/testkit.go]: Fixed

File: tests/realtikvtest/sessiontest/session_test.go (hunk 4 above)
Comment (hawkingrei) [reply tests/realtikvtest/sessiontest/session_test.go]: Done

File: tests/realtikvtest/sessiontest/session_test.go (hunk 3 above)
Comment (hawkingrei) [reply tests/realtikvtest/sessiontest/session_test.go]: Done

File: tests/realtikvtest/sessiontest/session_test.go (hunk 2 above)
Comment (hawkingrei) [reply tests/realtikvtest/sessiontest/session_test.go:1360]: Done

File: session/session_legacy_test.go [hunk 6]
```diff
@@ -2219,21 +2079,51 @@ func (s *testSessionSuite3) TestEnablePartition(c *C) {
...
+	result = tk.MustQuery("select @@group_concat_max_len;")
+	result.Check(testkit.Rows("4"))
+
+	_, err := tk.Exec("set @@group_concat_max_len = 18446744073709551616")
+	c.Assert(terror.ErrorEqual(err, variable.ErrWrongTypeForVar), IsTrue, Commentf("err %v", err))
+
+	// Test illegal type
+	_, err = tk.Exec("set @@group_concat_max_len='hello'")
+	c.Assert(terror.ErrorEqual(err, variable.ErrWrongTypeForVar), IsTrue, Commentf("err %v", err))
 }
```
Comment (tisonkun) [inline session/session_legacy_test.go:2127]:
```suggestion
```

File: tests/realtikvtest/sessiontest/session_test.go [hunk 7]
```diff
@@ -1021,3 +1040,674 @@ func TestDatabase(t *testing.T) {
...
+		tk.Session().RollbackTxn(context.Background())
+	}
+
+	// Also check GetSnapshotWithTS
+	snap := tk.Session().GetSnapshotWithTS(0)
+	val, err := snap.Get(context.Background(), k)
+	require.NoError(t, err)
+	require.Equal(t, []byte("v1"), val)
+}
+
```
Comment (tisonkun) [inline tests/realtikvtest/sessiontest/session_test.go:1415]:
```suggestion
```

**Review by lcwangchao — APPROVED** [review]

**Comment by tisonkun** [comment]
Great!

**Pull request merged**